modules/network/src/main/java/org/apache/ignite/internal/network/netty/ConnectionManager.java

```diff
@@ -92,6 +92,9 @@ public class ConnectionManager {
     /** Recovery descriptor provider. */
     private final RecoveryDescriptorProvider descriptorProvider = new DefaultRecoveryDescriptorProvider();
 
+    /** An action that is invoked before handshake completes. Only useful for tests. */
+    private volatile Runnable beforeHandshakeComplete = () -> {};
```

Comment:
I think we should avoid test code in prod code (except for getters, probably). Shall we maybe leverage the factory pattern to create different handshake handler?

Reply:
`RecoveryClientHandshakeManager` constructor currently requires 5 arguments. We can pass message factory directly, but we would still need to create a factory with 4 arguments which makes the factory sooo specific. I'm not sure whether it's the best way to go, but I'll try and we'll look at the result.

Reply:
Done. Does this look botter, WDYT?